### ISSUE TITLE
State inconsistency when component uses signal that decides whether it is rendered or not

### ISSUE THREAD

**Issue by mikeplus64** [ryansolid/solid#198]
In [solid-typefu-router5](https://github.com/mikeplus64/solid-typefu-router5) you have a `Link` component that renders a link given a router5 route name and the params (think `?id=1&page=2`) for the route.

If you then have a component for a user profile page somewhere with a link to statistics about that user, like `<Link to="userstats" params={{ id: getCurrentRoute().params.id }}>User statistics</Link>`, then all is well *until* you navigate off the user profile page, in which case Solid attempts to recompute the `href` for the user stats link because that `href` depended on `getCurrentRoute`, a signal representing the current route, which would have fired when you navigated off of the user profile page.

So even though the user profile page correctly unmounts, its effects/computations are run immediately before it is actually discarded. The effects don't know that the signal that caused their "owner" component to exist in the first place, has now changed, imminently sending that parent component to the bin.

## Minimal example

See https://codesandbox.io/s/vigilant-leftpad-bjl38?file=/src/index.tsx for an example.

Basically a signal is created which morally represents the route in the "real world" example I encountered. If the signal is `1` then you get a button with `1` as its label which sets the signal to `2`. If it's `2`, you get a button with `2` as its label which sets the signal to `1`. There's an effect in each that morally is the equivalent of e.g. a computation that sets a `href` or `className` or really anything.

I can "solve" it by wrapping the component immediately with `sample`, but that only stops the nearest "layer" of effects from running and completely breaks down on large components (like a user profile page).

Could component "existence" be treated as a a computation that occurs before any inner effects, and when a component is found to not exist or be scheduled for destruction, its computations discarded?

**Comment by mikeplus64** [comment]
A full example using routing:

https://codesandbox.io/s/eloquent-frog-gpvbf?file=/src/index.tsx

How it goes is like this:

1. User clicks "foo" from navbar
2. foo renders; user clicks "foo.child"
3. foo.child renders, queries an API (using its ID) for subchildren, and generates foo.child1.child2 links that look like `/foo/child1/${id1}/child2/${id2}`
4. User navigates to "bar" causing foo and foo.child to unmount
5. Error! The foo.child1.child2 links are regenerated due to the route changing (note their dependency on `route().params.id`) but that no longer exists at the `bar` route

**Comment by mikeplus64** [comment]
And another "minimal example". Interestingly the `Show` version works, but ternary operator and `Switch` don't.

**Comment by ryansolid** [comment]
Yeah this only happens with multiple switches and upon switching back and forth a multiple of times. I think it has to do with execution re-ordering during disposal. Which is going to take me some doing to work out. This bug I believe is in the reactive system cleanup code, so I'd need to make sure any change won't break anything else. This will probably take me a few days.

**Comment by mikeplus64** [comment]
> 
> 
> Yeah this only happens with multiple switches and upon switching back and forth a multiple of times.

Hm in these examples using either `createEffect` or `createMemo` inline, it happens immediately and with only one `Switch`

https://codesandbox.io/s/vigilant-leftpad-bjl38?file=/src/index.tsx (first example I posted but cleaned up slightly)

https://codesandbox.io/s/haunted-by-the-effects-of-the-past-switchmatchinline-componentcreatememo-dwbh0?file=/src/index.tsx:38-52 (variation using `createMemo`)

**Comment by ryansolid** [comment]
Yeah for an update as far as I can tell this is because on each cycle to optimize things I actually re-order the execution of subscribers. I thought it would have no side effects because of the push pull system I had cleverly looks ahead. But it only does so in terms of dependency tracking and not ownership. So to fix this I need to re-evaluate a few options. One is remove the disposal optimization (probably the easiest) but I need to guarantee this still can't happen. The other is have smarter tracking that if an owning context is marked as pending to defer evaluation until after it has resolved. I think that is ultimately the best and can fix other similar nasty problems but it will take more doing.

**Comment by mikeplus64** [comment]
> I consider this bug of the highest priority although I'm in the middle of setting the SSR release currently. So I need to figure out if I can break this apart in a reasonable way since I really also don't want to postpone the next set of API breaking changes since they are probably the last big ones before 1.0 and more and more people are using the library. That all being said I'm in the process of doing an international move this week so I don't see having time to do concentrated work on this during that time.

Yep fair enough. If I have time I'll try make a stab at it, if nothing else but for the Joy Of Learningtm =), hope the move goes smoothly!

**Comment by ryansolid** [comment]
> > Yeah this only happens with multiple switches and upon switching back and forth a multiple of times.
> 
> Hm in these examples using either `createEffect` or `createMemo` inline, it happens immediately and with only one `Switch`
> c
> https://codesandbox.io/s/vigilant-leftpad-bjl38?file=/src/index.tsx (first example I posted but cleaned up slightly)
> 
> https://codesandbox.io/s/haunted-by-the-effects-of-the-past-switchmatchinline-componentcreatememo-dwbh0?file=/src/index.tsx:38-52 (variation using `createMemo`)

These both do something a bit unusual in that they are returning functions which are creating their own context outside of the Switch. There is a function form but it doesn't work like that (it expects an argument, to differentiate from cases like this). So it is similar to having nested control flow. Other than those are wrapped in create memo which runs immediately and doesn't have this issue.

This brings up an interesting question. Hoisting into a component generally removes the issue. It's only leveraging something I need to allow but shouldn't be the pattern that allows this. Now I've been trying to decide where to auto apply memos. So far at the component level auto sets up guards against this. I use a trick in Context Providers to prevent this any level of nesting but it's a bit heavier. I was hoping that people generally weren't going to inline function declarations in the JSX as children and expect it to work. It does sort of work but it moves things to the parents scope. That's a different topic and I'm going to put my focus on the earlier example which highlights the reactivity issue.

There is a problem here for sure in the reactive system

**Comment by ryansolid** [comment]
I believe I've fixed this in the next release and more. I believe I've fixed the unrelated dependent case as well ie.. one signal for "show" and one for "data" and you `freeze` update but set data to `undefined` before setting `show` to false. This problem haunts most of these libraries and I believe I've solved it without a performance hit (just half a dozen more lines of code). I hope to release as soon as I'm done testing.
